sparse/impl/KokkosSparse_mdf_impl.hpp

```diff
@@ -161,6 +161,121 @@ struct MDF_discarded_fill_norm {
+  using static_crs_graph_type = typename crs_matrix_type::StaticCrsGraphType;
+  using col_ind_type =
+      typename static_crs_graph_type::entries_type::non_const_type;
+  using values_type  = typename crs_matrix_type::values_type::non_const_type;
+  using size_type    = typename crs_matrix_type::size_type;
+  using ordinal_type = typename crs_matrix_type::ordinal_type;
+  using scalar_type  = typename crs_matrix_type::value_type;
+  using KAS          = typename Kokkos::ArithTraits<scalar_type>;
+
+  const scalar_type zero = KAS::zero();
+
+  crs_matrix_type A, At;
+  ordinal_type factorization_step;
+  col_ind_type permutation;
+  col_ind_type update_list;
+
+  values_type discarded_fill;
+  col_ind_type deficiency;
+  int verbosity;
+
+  MDF_selective_discarded_fill_norm(crs_matrix_type A_, crs_matrix_type At_,
+                                    ordinal_type factorization_step_,
+                                    col_ind_type permutation_,
+                                    col_ind_type update_list_,
+                                    values_type discarded_fill_,
+                                    col_ind_type deficiency_, int verbosity_)
+      : A(A_),
+        At(At_),
+        factorization_step(factorization_step_),
+        permutation(permutation_),
+        update_list(update_list_),
+        discarded_fill(discarded_fill_),
+        deficiency(deficiency_),
+        verbosity(verbosity_){};
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const ordinal_type i) const {
+    ordinal_type rowIdx      = permutation(update_list(i));
+    scalar_type discard_norm = zero, diag_val = zero;
+    bool entryIsDiscarded       = true;
+    ordinal_type numFillEntries = 0;
+    for (size_type alphaIdx = At.graph.row_map(rowIdx);
+         alphaIdx < At.graph.row_map(rowIdx + 1); ++alphaIdx) {
+      ordinal_type fillRowIdx = At.graph.entries(alphaIdx);
+      bool row_not_eliminated = true;
+      for (ordinal_type stepIdx = 0; stepIdx < factorization_step; ++stepIdx) {
+        if (fillRowIdx == permutation(stepIdx)) {
+          row_not_eliminated = false;
+        }
+      }
+
+      if (fillRowIdx != rowIdx && row_not_eliminated) {
+        for (size_type betaIdx = A.graph.row_map(rowIdx);
+             betaIdx < A.graph.row_map(rowIdx + 1); ++betaIdx) {
+          ordinal_type fillColIdx = A.graph.entries(betaIdx);
+          bool col_not_eliminated = true;
+          for (ordinal_type stepIdx = 0; stepIdx < factorization_step;
+               ++stepIdx) {
+            if (fillColIdx == permutation(stepIdx)) {
+              col_not_eliminated = false;
+            }
+          }
+
+          if (fillColIdx != rowIdx && col_not_eliminated) {
+            entryIsDiscarded = true;
+            for (size_type entryIdx = A.graph.row_map(fillRowIdx);
+                 entryIdx < A.graph.row_map(fillRowIdx + 1); ++entryIdx) {
+              if (A.graph.entries(entryIdx) == fillColIdx) {
+                entryIsDiscarded = false;
```

Comment:
Same as above, although this loop only runs for as many entries as there are in a row